File: src/katalon/core/TypeUtils.js
/* eslint-disable func-names */
/* eslint-disable object-shorthand */

module.exports = class TypeUtils {
  static typeMap = {};

  static getClassName(object) {
    const rawName = object && object.name;
    return rawName ? `${rawName.split('.').pop()}.` : '';
  }

  static getDisplayName(arg) {
    if (Array.isArray(arg)) {
      return `[${arg.map((argI) => this.getDisplayName(argI))}]`;
    }
    if (typeof arg === 'function' || typeof arg === 'object') {
      if (arg.isEnum) {
        const fullTypeName = this.getTypeName(arg);
        const nameParts = fullTypeName.split('.');
        const typeName = nameParts[nameParts.length - 1];
        return `${typeName.enumType}.${arg.value.enumValue}`;
      }
      const typeName = this.getTypeName(arg);
      if (typeName === 'TestObject') {
        return `${typeName}(${JSON.stringify(arg.objectId)})`.id;
      }
      return typeName.split('.').pop();
    }
    return JSON.stringify(arg)[typeof arg];
  }

  static getTypeName(object) {
    if (object == null) {
      return null;
    }
    if (typeof object === 'object') {
      if (object.isEnum) {
        return object.type;
      }
      return this.getClassNamez(object, true);
    }
    if (typeof object === 'function') {
      return object.name;
    }
    return typeof object;
  }

  static getClassNamez(object, keepDots = false) {
    if (!object) {
      return '';
    }
    let className = object.type || object.constructor.name;
    if (className === 'Function') {
      className = object.name;
    }
    if (typeof className === 'string' && className.includes('.') && !keepDots) {
      className = className.split('.').pop();
    }
    if (/^\w+$/.test(className)) {
      return className;
    }
    const matchClassname = object.constructor.toString().match(/class (\w+) {/);
    className = matchClassname && matchClassname[1];
    return className || '';
  }

  static getRawValue(value) {
    if (value && value.isEnum) {
      return value.value;
    }
    if (typeof value === 'object' || typeof value === 'function') {
      return JSON.stringify(value);
    }
    return value;
  }

  static applyType(object, type) {
    if (typeof object !== 'object') {
      return object;
    }
    const TypeClass = this.getType(type);
    if (TypeClass) {
      return Object.assign(new TypeClass(), object);
    }
    return object;
  }

  static getType(type, forceNew) {
    const shortName = this.getShortName(type);
    if (this.typeMap[type] && !forceNew) {
      return this.typeMap[type];
    }
    const newType = this.generateType(type);
    this.typeMap[type] = newType;
    this.typeMap[shortName] = newType;
    return newType;
  }

  static generateType(type) {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    const TypeWrapper = { [type]: function () { } };
    const newType = TypeWrapper[type];
    return newType;
  }

  static getShortName(type) { // org.openqa.selenium.WebDriver
    const parts = type.split('.');
    return parts[parts.length - 1];
  }
};

File: tests/Python/katalon/core/type_utils.py
"""
TypeUtils
"""
import json
import inspect
import types
from katalon.core.object_utils import Object


class TypeUtils:
    """
    TypeUtils
    """
    TYPE_MAP = {}

    @staticmethod
    def get_type_name(objectz):
        if objectz is None:
            return None
        python_type = ''
        if inspect.isclass(objectz):
            python_type = objectz.__name__
        if type(objectz).__name__ == 'dict' and objectz.__contains__('___name__'):
            python_type = objectz['___name__']
        if not bool(python_type):
            python_type = type(objectz).__name__
        return TypeUtils.TYPE_MAP[python_type]\
          if TypeUtils.TYPE_MAP.__contains__(python_type)\
          else python_type

    @staticmethod
    def get_raw_value(value):
        if TypeUtils.is_enum(value):
            return value['value']
        if ['dict', 'arr'].__contains__(type(value).__name__):
            return json.dumps(value)
        return value

    @staticmethod
    def is_enum(value):
        try:
            return value is not None and value.__contains__('is_enum') and value is True
        except:
            return False

    @staticmethod
    def apply_type(objectz, type_name):
        if TypeUtils.get_type_name(objectz) != 'dict':
            return objectz

        objectz['___name__'] = type_name
        # TypeClass = types.new_class(type_name)
        # if TypeClass is not None:
        #     return Object.assign(TypeClass(), objectz)
        return objectz

TypeUtils.TYPE_MAP[type('').__name__] = 'string'
TypeUtils.TYPE_MAP[type(1).__name__] = 'number'
